File: 백준/Gold/7576. 토마토/토마토.java
import java.io.*;
import java.util.*;

public class Main {

    static int[][] tomato;
    static int[][] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int n = Integer.parseInt(input[0]);
        int m = Integer.parseInt(input[1]);

        tomato = new int[m][n];
        for (int i = 0; i < m; i++) {
            String[] line = br.readLine().split(" ");
            for (int j = 0; j < n; j++) {
                tomato[i][j] = Integer.parseInt(line[j]);
            }
        }

        visited = new int[m][n];
        bfs(n, m);
        int max = 0;
        boolean notTomato = false;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (visited[i][j] == 0) {
                    notTomato = true;
                }
                max = Math.max(max, visited[i][j]);
            }
        }
        if (notTomato) {
            System.out.println(-1);
        }
        else {
            System.out.println(max - 1);
        }
    }

    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};

    static void bfs(int n, int m) {
        Queue<int[]> q = new LinkedList<>();

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                // 익은 토마토 q에 넣기
                if (tomato[i][j] == 1) {
                    q.offer(new int[]{i, j});
                    visited[i][j] = 1;
                }
                // 빈 칸도 방문한 척
                if (tomato[i][j] == -1) {
                    visited[i][j] = 1;
                }
            }
        }

        while (!q.isEmpty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                int[] temp = q.poll();
                int r = temp[0];
                int c = temp[1];

                for (int d = 0; d < 4; d++) {
                    int nr = r + dr[d];
                    int nc = c + dc[d];

                    // 범위 체크
                    if (0 <= nr && nr < m && 0 <= nc && nc < n) {
                        if (visited[nr][nc] == 0) {
                            visited[nr][nc] = visited[r][c] + 1;
                            q.offer(new int[]{nr, nc});
                        }
                    }
                }
            }
        }
    }
}

File: 백준/Gold/7569. 토마토/토마토.py
from collections import deque

# m: 가로, n: 세로, h: 상자 수
m, n, h = map(int, input().split())
tomato = [[list(map(int, input().split())) for _ in range(n)] for _ in range(h)]

# 위 아래 왼 오 앞 뒤
dy = [0, 0, 0, 0, -1, 1]
dx = [0, 0, -1, 1, 0, 0]
dz = [1, -1, 0, 0, 0, 0]

def bfs(q, tomato):

    while q:
        length = len(q)

        for _ in range(length):
            z, y, x = q.popleft()

            for d in range(6):
                nz = z + dz[d]
                ny = y + dy[d]
                nx = x + dx[d]

                # 범위 체크
                if 0 <= ny < n and 0 <= nx < m and 0 <= nz < h:
                    # 안 익은 토마토인지 확인
                    if tomato[nz][ny][nx] == 0:
                        q.append((nz, ny, nx))
                        tomato[nz][ny][nx] = tomato[z][y][x] + 1


q = deque([])

# 1: 익은 토마토, 0: 안 익은 토마토, -1: 토마토 없음
for z in range(h):
    for y in range(n):
        for x in range(m):
            # q에 익은 토마토 전부 넣기
            if tomato[z][y][x] == 1:
                q.append((z, y, x))

bfs(q, tomato)
day = 0
all = True
for z in range(h):
    for y in range(n):
        for x in range(m):
            # 하나라도 안 익었으면 -1
            if tomato[z][y][x] == 0:
                all = False
            else:
                day = max(day, tomato[z][y][x])

if all:
    # 익은 토마토 1부터 더하기 시작했으니까
    print(day - 1)
else:
    print(-1)